Modernize some loops in torch

r-barnes opened the pull request:
Differential Revision: D29716813

malfet commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -579,10 +579,9 @@ void GraphTask::exec_post_processing() {
-    // NOLINTNEXTLINE(modernize-loop-convert)
-    for (size_t i = 0; i < final_callbacks_.size(); ++i) {
+    for (const auto& final_callback : final_callbacks_) {
       cb_lock.unlock();
-      final_callbacks_[i]();
+      final_callback();
       cb_lock.lock();
     }
```
See comment above, range loops could not be used here because callbacks can get mutated, which I guess causes CI failures

malfet reviewed (changes requested):
Please revert changes to `GraphTask::exec_post_processing` per comment above

malfet commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -579,8 +579,7 @@ void GraphTask::exec_post_processing() {
     // added in between callback calls, so iterators may become invalidated.
-    // NOLINTNEXTLINE(modernize-loop-convert)
-    for (size_t i = 0; i < final_callbacks_.size(); ++i) {
+    for (const auto i : c10::irange(final_callbacks_.size())) {
```
`c10::irange()` captures upper boundary at the start of the loop, but here it can change (for example increase, as tested in `TestAutograd.test_callback_adds_callback`), so any form of range loop is really no go. Perhaps in order to get rid of modernize-loop warning, a `while` loop could have been used here

r-barnes committed:
Modernize some loops in torch (#61737)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/61737

Test Plan: Sandcastle

Differential Revision: D29716813

fbshipit-source-id: fc84ce82257a14215aba2da7da59a5db375c2006
